### DISCUSSION TITLE
help me with the Standard API integration.

### DISCUSSION THREAD

**Discussion by hitkoch** [YetiForceCompany/YetiForce#33] [How to] [answered]
🐞 bug report

Web service APP = OK
![image](https://github.com/user-attachments/assets/0f3c3d1d-f6ad-4420-ae9d-b861bd626193)

Web service USERS = OK
![image](https://github.com/user-attachments/assets/6752fb6c-beb9-4d3d-952b-1a4277effed5)

POSTMAN =  ERROR
![image](https://github.com/user-attachments/assets/5f847c7d-e208-48d9-8153-7a3beecb8249)

I tested several endpoints, followed all the steps in some topics here, read the documentation several times.
I simulated the swagger.

But I can't make it work in my environment.
Can anyone help me with this task?
I have few dollars to pay for the service, but it will be paid.  👍

Thanks.

**Comment by mariuszkrzaczkowski** [comment]
you don't have rewrite rules set

https://github.com/YetiForceCompany/YetiForceCRM/blob/de296d83c1d36bf0d92303efa9706ded67549aa8/tests/setup/nginx/yetiforce.conf#L84-L90

**Reply by deepti4666** [reply]
Where do these rewrite rules go?

**Comment by deepti4666** [comment]
I keep getting a 401 Unauthorized
![image](https://github.com/user-attachments/assets/8fd20fd6-6a02-4e00-aa33-edb1bef8d2ea)

**Comment by rskrzypczak** [comment]
Just take a look here https://doc.yetiforce.com/en/developer-guides/api/
everything is described step by step

**Comment by adriangalbincea** [comment]
@rskrzypczak I have a similar issue. 
I created the API app/key, then I created a new user(specially for API), then WebService user with permissions based on user, only with Pass auth.
I have the rewrite rules. I did the installation as per your full guide. I did not miss a step.
After I setup the headers in postman, I still get an 401 error.
Below is my current configuration
![ice_screenshot_20250411-150857](https://github.com/user-attachments/assets/e1562fc9-8f81-4693-9156-0440b1c6c97b)

![ConfReport(1)](https://github.com/user-attachments/assets/255d694a-ed9b-491e-89e8-4c69c4b1ba5a)

**Reply by adriangalbincea** [reply]
I even tried to create the API on your demo and I get the same error.
Do I need to setup a password when I create Webservices user or I can leave it blank and I will authenticate with the user's email and password?

**Reply by adriangalbincea** [reply]
@deepti4666 did you manage to connect to the API in the end?

**Reply by adriangalbincea** [reply]
Here is the detailed error:
{
    "status": 0,
    "error": {
        "message": "Web service - Applications: Unauthorized",
        "code": 401,
        "file": "api/webservice/Core/Auth/Basic.php",
        "line": 25,
        "backtrace": "#0 api/webservice/Controller.php:185  >>  Api\\Core\\Exception->showError()\n#1 webservice.php:30  >>  Api\\Controller->handleError()\n#2 public_html/webservice.php:13  >>  require('webservice.php')"
    }
}

**Reply by rskrzypczak** [reply]
Have you checked this? https://doc.yetiforce.com/en/developer-guides/api/

**Reply by adriangalbincea** [reply]
I checked everything before I decided to bother you. The guide is almost straightforward.
1. I created a special user
2. I created an API Key. I had to setup a password, will that password be used anywhere or just the generated API key afterwards?
3. I created and API login in WebserviceStandard  under the API "server" above. Not sure why is called server when it should be call maybe Application

I checked the rewrite rules. I even tried to run postman with the link without rewrite and I still get the same error

I looked up the function that throws the error; and I can see that my error is not related to an incorrect password or invalid API key...
Is there an extra setting that must be done before the API to be used?

`        public function authenticate(string $realm): bool
        {
                if (!isset($_SERVER['PHP_AUTH_USER'])) {
                        $this->api->response->addHeader('WWW-Authenticate', 'Basic realm="' . $realm . '"');
                        throw new \Api\Core\Exception('Web service - Applications: Unauthorized', 401);
                }
                if (!$this->api->app || !$this->validatePwd($_SERVER['PHP_AUTH_PW'])) {
                        $this->api->response->addHeader('WWW-Authenticate', 'Basic realm="' . $realm . '"');
                        throw new \Api\Core\Exception('Web service - Applications: Wrong Credentials', 401);
                }
                $apiKey = $this->api->request->getHeaders()['x-api-key'] ?? null;
                if (!$apiKey || $apiKey !== \App\Encryption::getInstance()->decrypt($this->api->app['api_key'])) {
                        throw new \Api\Core\Exception('Invalid api key', 401);
                }

                return true;
        }`

**Reply by adriangalbincea** [reply]
Here is the Error from the Log file:

`============ Error exception ====== 2025-04-11 19:46:46 ======
code: 401 | message: Web service - Applications: Unauthorized
file: /home/yfprod/html/api/webservice/Core/Auth/Basic.php (25)
============ stacktrace:
#0 /home/yfprod/html/api/webservice/Core/Auth.php(45): Api\Core\Auth\Basic->authenticate('YetiForceApi')
#1 /home/yfprod/html/api/webservice/Controller.php(106): Api\Core\Auth::init(Object(Api\Controller))
#2 /home/yfprod/html/webservice.php(23): Api\Controller->preProcess()
#3 /home/yfprod/html/public_html/webservice.php(13): require('/home/yfprod/ht...')
#4 {main}
============ Request fd641f46 ======  2025-04-11 19:46:46  ======
REQUEST_METHOD: POST
REQUEST_URI: /webservice/WebserviceStandard/Users/Login
QUERY_STRING: _container=WebserviceStandard&module=Users&action=Login
PATH_INFO:
IP: 84.113.79.100
----------- Headers -----------
content-length: 80
connection: keep-alive
accept-encoding: gzip, deflate, br
host: force.wiseserve.net
postman-token: c769f766-848e-4386-8ed7-fac2906976ce
cache-control: no-cache
accept: */*
user-agent: PostmanRuntime/7.43.3
content-type: application/json
x-api-key: uGCNXKbUhrB3Asq1gFNQ7cePsBsktKFP
----------- Request data -----------
Array
(
    [_container] => WebserviceStandard
    [module] => Users
    [action] => Login
)

----------- _GET -----------
Array
(
    [_container] => WebserviceStandard
    [module] => Users
    [action] => Login
)

----------- Request payload -----------
{
    "userName": "no-reply@wiseserve.net",
    "password": "Adsaf234fdfas"
}`

**Reply by adriangalbincea** [reply]
The error is the same regardless of the credentials I add to the Body...

**Reply by hitkoch** [reply]
I studied so much that I'm already an expert in making the API work, if you need I can help you.

**Reply by adriangalbincea** [reply]
@hitkoch I would be grateful. I setup so many APIs until now but I cannot seem to understand what is wrong and where... What do you need from me?

**Reply by hitkoch** [reply, marked as answer]
I understand you, I spent more than 15 days to make it work, I even changed servers, I installed Yeti several times.
One momment, i send many prints, for help you.

**Reply by adriangalbincea** [reply]
Jeez, in the guide, everything looks so straightforward... Thanks. I am curious where might be the problem

**Reply by hitkoch** [reply]
1 - Go try Generate TOKEN, for next steps.

![image](https://github.com/user-attachments/assets/c83efc25-37a9-4e37-8804-58a57497f38d)

![image](https://github.com/user-attachments/assets/31bb1835-02c6-4d83-bfc4-d2a64e49c1c1)

![image](https://github.com/user-attachments/assets/9fff0c1e-d768-40eb-a887-f835688515bb)

This is where I went wrong, I entered the access user's password, but it is not the yetiforce access password, but rather the access password created there in the webservice user.
![image](https://github.com/user-attachments/assets/60137487-9eee-4200-b18e-ba594fb1b118)

Please check that you have all this information correct.

**Reply by adriangalbincea** [reply]
You are a STAR!!! It worked in this way. Now with that generated token I can move forward and GET and POST?

**Reply by hitkoch** [reply]
YES, YES, YES - with TOKEN go for others endpoints, GET, PUT, POST.

**Reply by hitkoch** [reply]
![image](https://github.com/user-attachments/assets/aebd3d04-2fa9-4b1a-ace9-38e0719bda2b)

**Reply by adriangalbincea** [reply]
I managed to list all the Contacts so it works. Can I list with Standard tickets too or that is part of the premium? I got permission denied

**Reply by hitkoch** [reply]
Hello Adrian, please send me print url.

**Reply by adriangalbincea** [reply]
Hi Ronald. I found the module name. It is called HelpDesk and not Tickets like the Module name in the GUI. I managed to create a PS script that pulls the token and then uses the token in another command to list/create tickets... Many many thanks for your help on this. With this achievement, I can look into migrating the current PSA into Yeti.
